### PR TITLE
12 mining service communication

### LINKED ISSUE #8: Accept transactions from the frontend and pass it to mining service
**Description:**

- When a user makes a payment, the transaction should be passed on to the blockchain to be added.

**Acceptance Criteria:**

- [x] The transactions are verified before being passed on to the blockchain.
- [x] The transactions are successfully added or denied on failures.
- [x] The front-end is notified of the results.

### LINKED ISSUE #12: Miner service to facilitate mining operations
**Description:**

The middleware will facilitate all communication between the miners and reward the appropriate miner on a first-come basis. 

- Initiate a docker container for the service
- Implement the proper endpoints required for the project
- Write unit tests and integration tests for all functions and endpoints

**Acceptance Criteria**
- [x] Mining works correctly, users are rewarded when mining works, a block is added to the blockchain when proofOfWork is computed

### PR DESCRIPTION
**Pull request by abhisachdev17**
## Summary of Changes
- Mining Pool Object which contains a pool of transactions
- One transaction is picked from the pool to be sent for Mining
- The transaction is only sent when the ready_to_mine condition variable with help of send is set.
- The ready_to_mine is set when the miners send back a valid proof of work. 

TLDR;  lots of OS stuff, let me know if you have any doubts on how things work.

### CODE REVIEW

File: backend/services/transactions/endpoints.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Transactions API
```
Comment (ZackHolmberg) [inline backend/services/transactions/endpoints.md]: Shouldnt have transaction service stuff in this PR

File: backend/services/transactions/src/routes.py [hunk 2]
```diff
@@ -1,5 +1,125 @@
 from src import app
```
Comment (ZackHolmberg) [inline backend/services/transactions/src/routes.py]: Shouldnt have transaction service stuff in this PR

File: backend/services/transactions/tests/functional/test_routes.py [hunk 3]
```diff
@@ -1,5 +1,11 @@
 import pytest
 from src import app
+from uuid import uuid4
```
Comment (ZackHolmberg) [inline backend/services/transactions/tests/functional/test_routes.py]: Shouldnt have transaction service stuff in this PR (This will be the last one. Just look through and remove every under services/transactions

File: backend/services/mining/src/MiningPool.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import threading
+import time
+
+class MiningPool:
+    DEQUEUE_TIME = 5
+
+    def __init__ (self, receiver, ready_to_mine):
+        self._pool = []
+        self._lock = threading.Lock()
+        self._mining_thread = threading.Thread(target=self.sendToMine, kwargs={'self': self, 'receiver': receiver})
```
Comment (ZackHolmberg) [inline backend/services/mining/src/MiningPool.py]: All lock stuff can be removed as per our conversation. I'd hazard a guess that clients and the service can push/pop to the existing list and things will be fine. However, to ensure this, we can use a Queue instead of a list.

**Review by ZackHolmberg — CHANGES_REQUESTED** [review]

**Comment by akshaysharma21** [comment]
We should probably replace "app.run()" to "socketio.run(app)" in run.py for mining. This method encapsulates the app.run() functionality and although both work, the latter is the recommended way to initialize the flask-socketio server because of better support for websockets. (According to the flask-socket,io docs)
